app/src/main/kotlin/com/mapbox/vision/teaser/MainActivity.kt

```diff
@@ -1,55 +1,531 @@
+
+        signSize = dpToPx(64f).toInt()
+        lineHeight = dpToPx(40f).toInt()
+        margin = dpToPx(8f).toInt()
+
+        back.setOnClickListener { onBackClick() }
+        segm_container.setOnClickListener { setVisionMode(VisionMode.Segmentation) }
+        sign_detection_container.setOnClickListener { setVisionMode(VisionMode.Classification) }
+        det_container.setOnClickListener { setVisionMode(VisionMode.Detection) }
+        distance_container.setOnClickListener { setVisionMode(VisionMode.Safety) }
+        safety_mode_container.hide()
+        line_detection_container.setOnClickListener { setVisionMode(VisionMode.Lanes) }
+
+        initRootLongTap()
+        fps_performance_view.hide()
+
+        initArNavigationButton()
+        initReplayModeButton()
+        tryToInitVisionManager()
+    }
+
+    private fun initRootLongTap() {
+        root.setOnLongClickListener {
+            if (fps_performance_view.visibility == View.GONE) {
+                fps_performance_view.show()
+            } else {
+                fps_performance_view.hide()
+            }
+            return@setOnLongClickListener true
+        }
+    }
+
+    private fun initArNavigationButton() {
+        ar_navigation_button_container.setOnClickListener {
+            when (visionManagerMode) {
+                Camera -> startActivity(Intent(this@MainActivity, ArMapActivity::class.java))
+                Replay -> startArSession()
+            }
+        }
+    }
+
+    private fun initReplayModeButton() {
+        replay_mode_button_container.setOnClickListener {
+            showReplayModeFragment()
+        }
+    }
+
+    private fun tryToInitVisionManager() {
+        if (isPermissionsGranted && !visionManagerWasInit) {
+            visionManagerWasInit = when (visionManagerMode) {
+                Camera -> initVisionManagerCamera(vision_view)
+                Replay -> initVisionManagerReplay(vision_view, sessionPath)
+            }
+        }
+    }
+
+    private fun stopVisionManager() {
+        if (isPermissionsGranted && visionManagerWasInit) {
+            when (visionManagerMode) {
+                Camera -> destroyVisionManagerCamera()
+                Replay -> destroyVisionManagerReplay()
+            }
+            visionManagerWasInit = false
+        }
+    }
+
+    override fun onResume() {
+        super.onResume()
+        tryToInitVisionManager()
+        vision_view.onResume()
+    }
+
+    override fun onPause() {
+        super.onPause()
+        stopVisionManager()
+        vision_view.onPause()
+        soundsPlayer.stop()
+    }
+
+    private fun onBackClick() {
+        soundsPlayer.stop()
+        dashboard_container.show()
+        hideLineDetectionContainer()
+        hideSignsContainer()
+        safety_mode_container.hide()
+        back.hide()
+    }
+
+    private fun drawSigns(signsValueUis: List<UiSign>) {
+        sign_info_container.removeAllViews()
+        for (signValue in signsValueUis) {
```

Comment:
```suggestion
        for (signValueUi in signsValueUis) {
```

Reply:
Maybe even `uiSign` and `uiSigns` for the list since the type name is `UiSign`

Reply:
fixed